src/tree/v2/providers/CompatibleBranchDataProvider.ts

```diff
@@ -0,0 +1,74 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { AzExtParentTreeItem, AzExtTreeDataProvider, AzExtTreeItem, IFindTreeItemContext, ISubscriptionContext, ITreeItemPickerContext } from '@microsoft/vscode-azext-utils';
+import { AppResource, AppResourceResolver } from '@microsoft/vscode-azext-utils/hostapi';
+import * as vscode from 'vscode';
+import { ApplicationResource, BranchDataProvider, ResourceModelBase } from '../../../api/v2/v2AzureResourcesApi';
+import { createSubscriptionContext } from '../../../utils/v2/credentialsUtils';
+
+export class CompatibleBranchDataProvider<TResource extends ApplicationResource, TModel extends AzExtTreeItem & ResourceModelBase> extends AzExtTreeDataProvider implements BranchDataProvider<TResource, TModel> {
+    private readonly overrideOnDidChangeTreeDataEmitter = new vscode.EventEmitter<TModel | undefined>();
+
+    public constructor(private readonly resolver: AppResourceResolver, loadMoreCommandId: string) {
+        super(undefined as unknown as AzExtParentTreeItem, loadMoreCommandId);
+    }
+
+    //#region TreeDataProvider
+
+    public override readonly onDidChangeTreeData = this.overrideOnDidChangeTreeDataEmitter.event;
+
+    // @ts-expect-error `getParent` is not meant to be defined by `BranchDataProvider`s but is already defined by `AzExtTreeDataProvider`
+    public override getParent(_treeItem: TModel): Promise<TModel> {
+        throw new Error('Use the Resources extension API to do getParent');
+    }
+
+    public override getChildren(treeItem: TModel & AzExtParentTreeItem): Promise<TModel[]> {
+        // This method is redeclared to make TypeScript happier, but it's no more than a super call with extra casts
+        return super.getChildren(treeItem) as Promise<TModel[]>;
+    }
+
+    //#endregion
+
+    //#region BranchDataProvider
+
+    public async getResourceItem(element: TResource): Promise<TModel> {
+        const oldAppResource: AppResource = {
+            ...element,
+            type: element.type.type,
+            kind: element.type.kinds?.join(';'),
+        };
+        const subscriptionContext: ISubscriptionContext = createSubscriptionContext(element.subscription);
+
+        return await this.resolver.resolveResource(subscriptionContext, oldAppResource) as Promise<TModel>;

```

Comment:
This is going to return a `ResolvedAppResourceBase` which is not an `AzExtTreeItem` by itself. Currently we "combine" a `ResolvedAppResourceBase` with an `AppResourceTreeItem` using a `Proxy`.

I think we'll need to make a compat version of this and return it here, because just returning a `ResolvedAppResourceBase` doesn't have the necessary properties fro the `AzExtTreeDataProvider` to turn it into a `vscode.TreeItem`.

Reply:
Yeah. I think we'll have to evaluate what parts of `AzExtTreeItem` are needed and include them.